core/embed/rust/src/trezorhal/storage.rs

```diff
@@ -0,0 +1,330 @@
+pub fn init() {
+    unsafe {
+        ffi::storage_init(
+            Some(callback_wrapper),
+            ffi::HW_ENTROPY_DATA.as_ptr(),
+            ffi::HW_ENTROPY_DATA.len() as u16,
+        );
+    }
+}
+
+/// Set the PIN callback function.
+/// See documentation for `PinDelayCallback` for more information.
+pub fn set_pin_delay_callback(callback: PinDelayCallback) {
+    unsafe {
+        PIN_UI_CALLBACK = Some(callback);
+    }
+}
+
+/// Wipe storage.
+pub fn wipe() {
+    unsafe { ffi::storage_wipe() }
+}
+
+/// Check if storage is unlocked.
+pub fn is_unlocked() -> bool {
+    ffi::sectrue == unsafe { ffi::storage_is_unlocked() }
+}
+
+/// Lock storage.
+pub fn lock() {
+    unsafe { ffi::storage_lock() }
+}
+
+/// Unlock storage with PIN and optional external salt.
+/// Returns true if the PIN + salt combination is correct.
+pub fn unlock(pin: &str, salt: Option<&ExternalSalt>) -> bool {
+    let salt = salt.map(|s| s.as_ptr()).unwrap_or(ptr::null());
+    ffi::sectrue == unsafe { ffi::storage_unlock(pin.as_ptr() as *const _, pin.len(), salt) }
+}
+
+/// Change PIN and/or external salt.
+/// Returns true if the PIN + salt combination is correct and the change was
+/// successful.
+pub fn change_pin(
+    old_pin: &str,
+    new_pin: &str,
+    old_salt: Option<&ExternalSalt>,
+    new_salt: Option<&ExternalSalt>,
+) -> bool {
+    ffi::sectrue
+        == unsafe {
+            ffi::storage_change_pin(
+                old_pin.as_ptr() as *const _,
+                old_pin.len(),
+                new_pin.as_ptr() as *const _,
+                new_pin.len(),
+                old_salt.map(|s| s.as_ptr()).unwrap_or(ptr::null()),
+                new_salt.map(|s| s.as_ptr()).unwrap_or(ptr::null()),
+            )
+        }
+}
+
+/// Check if storage has PIN set.
+pub fn has_pin() -> bool {
+    ffi::sectrue == unsafe { ffi::storage_has_pin() }
+}
+
+/// Get remaining PIN attempts.
+pub fn get_pin_remaining() -> u32 {
+    unsafe { ffi::storage_get_pin_rem() }
+}
+
+pub fn ensure_not_wipe_pin(pin: &str) {
+    unsafe {
+        ffi::storage_ensure_not_wipe_code(pin.as_ptr(), pin.len());
+    }
+}
+
+/// Check if value for `appkey` exists.
+pub fn has(appkey: u16) -> bool {
+    ffi::sectrue == unsafe { ffi::storage_has(appkey) }
+}
+
+/// Get length of value stored for `appkey`.
+/// Returns an error if the value does not exist.
+pub fn get_length(appkey: u16) -> StorageResult<usize> {
+    let mut len = 0;
+    if ffi::sectrue == unsafe { ffi::storage_get(appkey, ptr::null_mut() as _, 0, &mut len) } {
+        Ok(len as usize)
+    } else {
+        Err(StorageError::ReadFailed)
+    }
+}
+
+/// Get value for `appkey` from storage.
+/// The provided `data` buffer must be long enough to hold the value. If the
+/// buffer is shorter than the stored value, an error is returned. Use
+/// `get_length` to find the size of the value beforehand. On success, the data
+/// is copied to `data`, and length of copied data is returned. Data previously
```

Comment:
please also update the comment ("length of copied data" -> subslice)

Reply:
https://github.com/trezor/trezor-firmware/commit/75587ee1b01dcd451c80a8142697dc5c6154965a